<a href="https://colab.research.google.com/github/JohnSRX/Machine-Learning-Portfolio/blob/Assosiation-Rules/Vehicle_AR_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder

In [ ]:
df=pd.read_csv('/content/CARS.csv')
df.head()

FORD,TOYOTA,CHEVEROLET
0  TOYOTA,FORD,CHEVEROLET,JEEP
1           TOYOTA,HONDA,DODGE
2  CHRYSLER,SUBARU,TOYOTA,FORD
3      SUBARU,HONDA,CHEVEROLET
4           TOYOTA,HONDA,DODGE

In [ ]:
df=pd.read_csv('/content/CARS.csv', sep=',')
df.head()

FORD,TOYOTA,CHEVEROLET
0  TOYOTA,FORD,CHEVEROLET,JEEP
1           TOYOTA,HONDA,DODGE
2  CHRYSLER,SUBARU,TOYOTA,FORD
3      SUBARU,HONDA,CHEVEROLET
4           TOYOTA,HONDA,DODGE

In [ ]:
df=pd.read_csv('/content/CARS.csv', names=['Vehicles'])
df.head() 

Vehicles
0       FORD,TOYOTA,CHEVEROLET
1  TOYOTA,FORD,CHEVEROLET,JEEP
2           TOYOTA,HONDA,DODGE
3  CHRYSLER,SUBARU,TOYOTA,FORD
4      SUBARU,HONDA,CHEVEROLET

In [ ]:
df.shape

(20, 1)

In [ ]:
data=list(df["Vehicles"].apply(lambda x:x.split(",")))
data

[['FORD', 'TOYOTA', 'CHEVEROLET'],
 ['TOYOTA', 'FORD', 'CHEVEROLET', 'JEEP'],
 ['TOYOTA', 'HONDA', 'DODGE'],
 ['CHRYSLER', 'SUBARU', 'TOYOTA', 'FORD'],
 ['SUBARU', 'HONDA', 'CHEVEROLET'],
 ['TOYOTA', 'HONDA', 'DODGE'],
 ['SUBARU', 'HONDA', 'JEEP'],
 ['SUBARU', 'TOYOTA', 'HONDA', 'CHEVEROLET'],
 ['CHRYSLER', 'SUBARU', 'TOYOTA', 'HONDA'],
 ['TOYOTA', 'FORD'],
 ['MAZDA', 'ACURA', 'CHEVEROLET', 'JEEP'],
 ['MAZDA', 'ACURA', 'CHEVEROLET', 'JEEP'],
 ['MAZDA', 'TESLA', 'DODGE'],
 ['TOYOTA', 'MAZDA', 'ACURA'],
 ['TOYOTA', 'TESLA', 'CHEVEROLET'],
 ['MAZDA', 'TESLA', 'JEEP'],
 ['TOYOTA', 'TESLA', 'DODGE'],
 ['TOYOTA', 'MAZDA', 'TESLA'],
 ['TOYOTA', 'MAZDA', 'TESLA'],
 ['HONDA', 'FORD', 'MAZDA', 'JEEP']]

In [ ]:
from mlxtend.preprocessing import TransactionEncoder
a=TransactionEncoder()
a_data=a.fit(data).transform(data)
df=pd.DataFrame(a_data, columns=a.columns_)

df=df.replace(False,0)
df=df.replace(True,1)

df

ACURA  CHEVEROLET  CHRYSLER  DODGE  FORD  HONDA  JEEP  MAZDA  SUBARU  \
0       0           1         0      0     1      0     0      0       0   
1       0           1         0      0     1      0     1      0       0   
2       0           0         0      1     0      1     0      0       0   
3       0           0         1      0     1      0     0      0       1   
4       0           1         0      0     0      1     0      0       1   
5       0           0         0      1     0      1     0      0       0   
6       0           0         0      0     0      1     1      0       1   
7       0           1         0      0     0      1     0      0       1   
8       0           0         1      0     0      1     0      0       1   
9       0           0         0      0     1      0     0      0       0   
10      1           1         0      0     0      0     1      1       0   
11      1           1         0      0     0      0     1      1       0   
12      0           0         0      1     0      0     0      1       0   
13      1           0         0      0     0      0     0      1       0   
14      0           1         0      0     0      0     0      0       0   
15      0           0         0      0     0      0     1      1       0   
16      0           0         0      1     0      0     0      0       0   
17      0           0         0      0     0      0     0      1       0   
18      0           0         0      0     0      0     0      1       0   
19      0           0         0      0     1      1     1      1       0   

    TESLA  TOYOTA  
0       0       1  
1       0       1  
2       0       1  
3       0       1  
4       0       0  
5       0       1  
6       0       0  
7       0       1  
8       0       1  
9       0       1  
10      0       0  
11      0       0  
12      1       0  
13      0       1  
14      1       1  
15      1       0  
16      1       1  
17      1       1  
18      1       1  
19      0       0

In [ ]:
df=apriori(df,min_support=0.2,use_colnames=True)
df

support              itemsets
0      0.35          (CHEVEROLET)
1      0.20               (DODGE)
2      0.25                (FORD)
3      0.35               (HONDA)
4      0.30                (JEEP)
5      0.40               (MAZDA)
6      0.25              (SUBARU)
7      0.30               (TESLA)
8      0.65              (TOYOTA)
9      0.20  (CHEVEROLET, TOYOTA)
10     0.20        (FORD, TOYOTA)
11     0.20       (SUBARU, HONDA)
12     0.20       (HONDA, TOYOTA)
13     0.20         (MAZDA, JEEP)
14     0.20        (MAZDA, TESLA)
15     0.20       (TOYOTA, TESLA)

In [ ]:
df_ar=association_rules(df,metric="confidence", min_threshold=0.6)
df_ar

antecedents consequents  antecedent support  consequent support  support  \
0      (FORD)    (TOYOTA)                0.25                0.65      0.2   
1    (SUBARU)     (HONDA)                0.25                0.35      0.2   
2      (JEEP)     (MAZDA)                0.30                0.40      0.2   
3     (TESLA)     (MAZDA)                0.30                0.40      0.2   
4     (TESLA)    (TOYOTA)                0.30                0.65      0.2   

   confidence      lift  leverage  conviction  
0    0.800000  1.230769    0.0375        1.75  
1    0.800000  2.285714    0.1125        3.25  
2    0.666667  1.666667    0.0800        1.80  
3    0.666667  1.666667    0.0800        1.80  
4    0.666667  1.025641    0.0050        1.05

If you look at the first set of row, you can see that Ford and Toyota have a CONFIDENCE of .80, This is because there is an 80% chance that someone who owns a ford, also owns a toyota as their second vehicle. If you looka at the Tesla and Toyota row, you can see that they 67% chance of having both vehicles.

In [ ]:
df_ar=association_rules(df,metric="lift", min_threshold=1)
df_ar

antecedents consequents  antecedent support  consequent support  support  \
0      (FORD)    (TOYOTA)                0.25                0.65      0.2   
1    (TOYOTA)      (FORD)                0.65                0.25      0.2   
2    (SUBARU)     (HONDA)                0.25                0.35      0.2   
3     (HONDA)    (SUBARU)                0.35                0.25      0.2   
4     (MAZDA)      (JEEP)                0.40                0.30      0.2   
5      (JEEP)     (MAZDA)                0.30                0.40      0.2   
6     (MAZDA)     (TESLA)                0.40                0.30      0.2   
7     (TESLA)     (MAZDA)                0.30                0.40      0.2   
8    (TOYOTA)     (TESLA)                0.65                0.30      0.2   
9     (TESLA)    (TOYOTA)                0.30                0.65      0.2   

   confidence      lift  leverage  conviction  
0    0.800000  1.230769    0.0375    1.750000  
1    0.307692  1.230769    0.0375    1.083333  
2    0.800000  2.285714    0.1125    3.250000  
3    0.571429  2.285714    0.1125    1.750000  
4    0.500000  1.666667    0.0800    1.400000  
5    0.666667  1.666667    0.0800    1.800000  
6    0.500000  1.666667    0.0800    1.400000  
7    0.666667  1.666667    0.0800    1.800000  
8    0.307692  1.025641    0.0050    1.011111  
9    0.666667  1.025641    0.0050    1.050000

If you look at the third row, you can see that Subaru and Honda have a LIFT of 2.285714, This is because there is a really positive coorilation between them. If you look at the ninth row, you can see that Tesla and Toyota have a LIFT of 1.025641. This means that they just have a slightly positive coorilation.

In [ ]:
df_ar=association_rules(df,metric="conviction", min_threshold=1.0)
df_ar


antecedents consequents  antecedent support  consequent support  support  \
0      (FORD)    (TOYOTA)                0.25                0.65      0.2   
1    (TOYOTA)      (FORD)                0.65                0.25      0.2   
2    (SUBARU)     (HONDA)                0.25                0.35      0.2   
3     (HONDA)    (SUBARU)                0.35                0.25      0.2   
4     (MAZDA)      (JEEP)                0.40                0.30      0.2   
5      (JEEP)     (MAZDA)                0.30                0.40      0.2   
6     (MAZDA)     (TESLA)                0.40                0.30      0.2   
7     (TESLA)     (MAZDA)                0.30                0.40      0.2   
8    (TOYOTA)     (TESLA)                0.65                0.30      0.2   
9     (TESLA)    (TOYOTA)                0.30                0.65      0.2   

   confidence      lift  leverage  conviction  
0    0.800000  1.230769    0.0375    1.750000  
1    0.307692  1.230769    0.0375    1.083333  
2    0.800000  2.285714    0.1125    3.250000  
3    0.571429  2.285714    0.1125    1.750000  
4    0.500000  1.666667    0.0800    1.400000  
5    0.666667  1.666667    0.0800    1.800000  
6    0.500000  1.666667    0.0800    1.400000  
7    0.666667  1.666667    0.0800    1.800000  
8    0.307692  1.025641    0.0050    1.011111  
9    0.666667  1.025641    0.0050    1.050000

The CONVICTION between Subaru and HONDA is 3.250000. This means that there is a very positive coorilation between the two. The CONVICTION between TOYOTA and TESLA is 1.011111. This means that there is a negative coorilation between the two. 

In [ ]:
df_ar=association_rules(df,metric="confidence", min_threshold=0.3)
df_ar

antecedents   consequents  antecedent support  consequent support  \
0   (CHEVEROLET)      (TOYOTA)                0.35                0.65   
1       (TOYOTA)  (CHEVEROLET)                0.65                0.35   
2         (FORD)      (TOYOTA)                0.25                0.65   
3       (TOYOTA)        (FORD)                0.65                0.25   
4       (SUBARU)       (HONDA)                0.25                0.35   
5        (HONDA)      (SUBARU)                0.35                0.25   
6        (HONDA)      (TOYOTA)                0.35                0.65   
7       (TOYOTA)       (HONDA)                0.65                0.35   
8        (MAZDA)        (JEEP)                0.40                0.30   
9         (JEEP)       (MAZDA)                0.30                0.40   
10       (MAZDA)       (TESLA)                0.40                0.30   
11       (TESLA)       (MAZDA)                0.30                0.40   
12      (TOYOTA)       (TESLA)                0.65                0.30   
13       (TESLA)      (TOYOTA)                0.30                0.65   

    support  confidence      lift  leverage  conviction  
0       0.2    0.571429  0.879121   -0.0275    0.816667  
1       0.2    0.307692  0.879121   -0.0275    0.938889  
2       0.2    0.800000  1.230769    0.0375    1.750000  
3       0.2    0.307692  1.230769    0.0375    1.083333  
4       0.2    0.800000  2.285714    0.1125    3.250000  
5       0.2    0.571429  2.285714    0.1125    1.750000  
6       0.2    0.571429  0.879121   -0.0275    0.816667  
7       0.2    0.307692  0.879121   -0.0275    0.938889  
8       0.2    0.500000  1.666667    0.0800    1.400000  
9       0.2    0.666667  1.666667    0.0800    1.800000  
10      0.2    0.500000  1.666667    0.0800    1.400000  
11      0.2    0.666667  1.666667    0.0800    1.800000  
12      0.2    0.307692  1.025641    0.0050    1.011111  
13      0.2    0.666667  1.025641    0.0050    1.050000

If you look at the fourth row, you can see that Subaru and Honda have a CONFIDENCE of .80, This is because there is an 80% chance that someone who owns a ford, also owns a toyota as their second vehicle. If you looka at the Tesla and Mazda row, you can see that they 67% chance of having both vehicles.

In [ ]:
df_ar=association_rules(df,metric="lift", min_threshold=0.8)
df_ar

antecedents   consequents  antecedent support  consequent support  \
0   (CHEVEROLET)      (TOYOTA)                0.35                0.65   
1       (TOYOTA)  (CHEVEROLET)                0.65                0.35   
2         (FORD)      (TOYOTA)                0.25                0.65   
3       (TOYOTA)        (FORD)                0.65                0.25   
4       (SUBARU)       (HONDA)                0.25                0.35   
5        (HONDA)      (SUBARU)                0.35                0.25   
6        (HONDA)      (TOYOTA)                0.35                0.65   
7       (TOYOTA)       (HONDA)                0.65                0.35   
8        (MAZDA)        (JEEP)                0.40                0.30   
9         (JEEP)       (MAZDA)                0.30                0.40   
10       (MAZDA)       (TESLA)                0.40                0.30   
11       (TESLA)       (MAZDA)                0.30                0.40   
12      (TOYOTA)       (TESLA)                0.65                0.30   
13       (TESLA)      (TOYOTA)                0.30                0.65   

    support  confidence      lift  leverage  conviction  
0       0.2    0.571429  0.879121   -0.0275    0.816667  
1       0.2    0.307692  0.879121   -0.0275    0.938889  
2       0.2    0.800000  1.230769    0.0375    1.750000  
3       0.2    0.307692  1.230769    0.0375    1.083333  
4       0.2    0.800000  2.285714    0.1125    3.250000  
5       0.2    0.571429  2.285714    0.1125    1.750000  
6       0.2    0.571429  0.879121   -0.0275    0.816667  
7       0.2    0.307692  0.879121   -0.0275    0.938889  
8       0.2    0.500000  1.666667    0.0800    1.400000  
9       0.2    0.666667  1.666667    0.0800    1.800000  
10      0.2    0.500000  1.666667    0.0800    1.400000  
11      0.2    0.666667  1.666667    0.0800    1.800000  
12      0.2    0.307692  1.025641    0.0050    1.011111  
13      0.2    0.666667  1.025641    0.0050    1.050000

If you look at the fourth row, you can see that Subaru and Honda have a LIFT of 2.285714, This is because there is a really positive coorilation between them. If you look at the ninth row, you can see that Honda and Toyota have a LIFT of 0.879121. This means that they just have a negative coorilation.

In [ ]:
df_ar=association_rules(df,metric="conviction", min_threshold=0.8)
df_ar

antecedents   consequents  antecedent support  consequent support  \
0   (CHEVEROLET)      (TOYOTA)                0.35                0.65   
1       (TOYOTA)  (CHEVEROLET)                0.65                0.35   
2         (FORD)      (TOYOTA)                0.25                0.65   
3       (TOYOTA)        (FORD)                0.65                0.25   
4       (SUBARU)       (HONDA)                0.25                0.35   
5        (HONDA)      (SUBARU)                0.35                0.25   
6        (HONDA)      (TOYOTA)                0.35                0.65   
7       (TOYOTA)       (HONDA)                0.65                0.35   
8        (MAZDA)        (JEEP)                0.40                0.30   
9         (JEEP)       (MAZDA)                0.30                0.40   
10       (MAZDA)       (TESLA)                0.40                0.30   
11       (TESLA)       (MAZDA)                0.30                0.40   
12      (TOYOTA)       (TESLA)                0.65                0.30   
13       (TESLA)      (TOYOTA)                0.30                0.65   

    support  confidence      lift  leverage  conviction  
0       0.2    0.571429  0.879121   -0.0275    0.816667  
1       0.2    0.307692  0.879121   -0.0275    0.938889  
2       0.2    0.800000  1.230769    0.0375    1.750000  
3       0.2    0.307692  1.230769    0.0375    1.083333  
4       0.2    0.800000  2.285714    0.1125    3.250000  
5       0.2    0.571429  2.285714    0.1125    1.750000  
6       0.2    0.571429  0.879121   -0.0275    0.816667  
7       0.2    0.307692  0.879121   -0.0275    0.938889  
8       0.2    0.500000  1.666667    0.0800    1.400000  
9       0.2    0.666667  1.666667    0.0800    1.800000  
10      0.2    0.500000  1.666667    0.0800    1.400000  
11      0.2    0.666667  1.666667    0.0800    1.800000  
12      0.2    0.307692  1.025641    0.0050    1.011111  
13      0.2    0.666667  1.025641    0.0050    1.050000

The CONVICTION between Subaru and HONDA is 3.250000. This means that there is a very positive coorilation between the two. The CONVICTION between TOYOTA and HONDA is 0.816667. This means that there is a negative coorilation between them. 